File: 拍名片/ScannedCard.swift
import Foundation

struct LabeledValue: Equatable, Identifiable {
    enum Kind: String, CaseIterable, Codable {
        case mobile
        case work
        case fax
        case home
        case other

        var displayName: String {
            switch self {
            case .mobile:
                return "手機"
            case .work:
                return "公司"
            case .fax:
                return "傳真"
            case .home:
                return "住家"
            case .other:
                return "其他"
            }
        }
    }

    let id: UUID
    var kind: Kind
    var value: String

    init(id: UUID = UUID(), kind: Kind, value: String) {
        self.id = id
        self.kind = kind
        self.value = value
    }

    var isEmpty: Bool {
        value.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
    }
}

struct ScannedCard: Equatable {
    var givenName = ""
    var familyName = ""
    var company = ""
    var jobTitle = ""
    var phoneNumbers: [LabeledValue] = []
    var emails: [LabeledValue] = []
    var address = ""

    var fullName: String {
        let name = [familyName, givenName]
            .filter { !$0.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty }
            .joined()

        if !name.isEmpty {
            return name
        }

        return [givenName, familyName]
            .filter { !$0.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty }
            .joined(separator: " ")
    }

    var displayName: String {
        let name = fullName.trimmingCharacters(in: .whitespacesAndNewlines)
        return name.isEmpty ? "未命名聯絡人" : name
    }

    var hasContent: Bool {
        ![
            givenName,
            familyName,
            company,
            jobTitle,
            address
        ].allSatisfy { $0.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty } ||
        phoneNumbers.contains(where: { !$0.isEmpty }) ||
        emails.contains(where: { !$0.isEmpty })
    }

    var primaryPhone: String {
        phoneNumbers.first(where: { !$0.isEmpty })?.value ?? ""
    }

    var primaryEmail: String {
        emails.first(where: { !$0.isEmpty })?.value ?? ""
    }

    mutating func normalized() {
        phoneNumbers = phoneNumbers.filter { !$0.isEmpty }
        emails = emails.filter { !$0.isEmpty }
    }
}

File: paimingpian-android/app/src/main/java/com/ifachang/paimingpian/model/ScannedCard.kt
package com.ifachang.paimingpian.model

data class LabeledValue(
    val kind: Kind,
    val value: String
) {
    enum class Kind(val displayName: String) {
        MOBILE("手機"),
        WORK("公司"),
        FAX("傳真"),
        HOME("住家"),
        OTHER("其他")
    }

    val isEmpty: Boolean
        get() = value.isBlank()
}

data class ScannedCard(
    val givenName: String = "",
    val familyName: String = "",
    val company: String = "",
    val jobTitle: String = "",
    val phoneNumbers: List<LabeledValue> = emptyList(),
    val emails: List<LabeledValue> = emptyList(),
    val address: String = ""
) {
    val fullName: String
        get() {
            val compactName = listOf(familyName, givenName).filter { it.isNotBlank() }.joinToString("")
            return compactName.ifBlank {
                listOf(givenName, familyName).filter { it.isNotBlank() }.joinToString(" ")
            }
        }

    val displayName: String
        get() = fullName.ifBlank { "未命名聯絡人" }

    val hasContent: Boolean
        get() = listOf(givenName, familyName, company, jobTitle, address).any { it.isNotBlank() } ||
            phoneNumbers.any { !it.isEmpty } ||
            emails.any { !it.isEmpty }

    fun normalized(): ScannedCard {
        return copy(
            phoneNumbers = phoneNumbers.filterNot { it.isEmpty },
            emails = emails.filterNot { it.isEmpty }
        )
    }
}

object DemoCards {
    val sample = ScannedCard(
        givenName = "Ivan",
        familyName = "Chang",
        company = "WoWo AI Commerce",
        jobTitle = "Founder",
        phoneNumbers = listOf(
            LabeledValue(LabeledValue.Kind.MOBILE, "+886 987 605 116")
        ),
        emails = listOf(
            LabeledValue(LabeledValue.Kind.WORK, "ifa1002@icloud.com")
        ),
        address = "台北市信義區松高路"
    )
}
